### ISSUE TITLE
Customizable graph style

### ISSUE THREAD

**Issue by jing-bi** [foambubble/foam#436]
It would be great that we can add a css file to modify the graph style.
The changeable style might be:
- color of the node
- thickness of the link
- highlight style when hovering the node
- force between nodes

**Comment by jmg-duarte** [comment]
This has been in the works, but the graph and its place in the project has been subject to discussion

**Comment by riccardoferretti** [comment]
To clarify @jmg-duarte, although we are debating how to organize the graph code, that doesn't stop us from implementing new features to the graph in the meantime.
I am actually very keen to add custom styling ASAP, there are a couple of considerations (where does the style sheet live, which properties to expose, ..) but nothing that is impacted by where the graph code comes from.

At the moment these are the style properties available:
- background-color
- font-size
- highlight-color
- node-color by node type
  - note
  - placeholder
  - unknown (I don't think this one is used)

I would expect in the future:
- color for tags (currently we don't show tags)
- color for other node types (based on e.g. a `type` property on the node)
- color for attachments (currently we don't show attachments)

**Comment by jmg-duarte** [comment]
I'm interested in addressing this as I think some aspects of the graph are currently _not optimal_.
I think it makes sense we take this on a step by step basis instead of trying to fit every possible feature into a single PR.

# Style File

## Format
### JavaScript
@riccardoferretti proposed to use `.js`, however I dislike the idea since it opens the possibility for code execution within what was supposed to be a style file

### CSS
CSS is usually the default for this kind of things, however, the styling we require is not more complex than some colors, and sizes.
It ends up being unnecessarily verbose and complex for new users.

### JSON
I my opinion, this is probably the most practical solution, we can simply load the file as an object, filter the required fields and fill in the missing ones.
It is also structured, exposing the hierarchy in styles, and being somewhat reflective of WYSIWYG/more intuitive since the hierarchy is laid out in front of the user.

## Location
Regarding the file location, I am not opposed to adding a `.foam` folder moving forward and deprecating the `.vscode/foam.json` file, it seems to be an overall better solution.

However, I suggest that the file placement be customizable, enabling the user to set the path to the file (or maybe even the folder). This would enable a global Foam config moving forward. (AFAIK `foam-core` is already moving in this direction)

## Naming
If the file path is customizable, this is a non issue. Regardless I propose `graph.json` as the default name.

## Structure 
The current structure is as follows:
```js
const style = {
  background: "",
  fontSize: 10,
  highlightedForeground: "",
  node: {
    note: "",
    nonExistingNote: "",
    unknown: ""
  }
};
```
We can just reuse the existing structure and expand as needed.

## Notes
I do not think that the graph config and other _VSCode only_ related files should be known by `foam-core`.
Exactly because of its name. While I know that below the surface, Foam structure _is_ a graph, I do not think its visualization should be part of the core, not until other tools make use of it.

**Comment by riccardoferretti** [comment]
Thanks for the thorough write up, here are my comments:

#### JSON file format

Sounds good to me.

#### File location

We already have a global config file (`~/.foam/config.json`) that is read by Foam, see `foam-core/src/config.ts:57`.
Having a customizable file location doesn't solve the problem of where you read the configuration for that path in the first place, I reckon you were thinking VsCode settings file?
I don't believe this needs to be customizable on the first iteration, feels like a pretty specific use case, as currently you can have project specific, or user specific config - which should cover most situations. 
The only remaining case is if you want to share config across projects (so you can't use the project config), but not all the same (so you can't use the user config)).

I would actually keep things simple and have the config in the VsCode settings file, for the following reasons:
- we keep one way of doing configuration
- single source of truth
- we already have some config migration to do, doing it for the graph too won't be extra work
- we can leverage the project/user level configuration scoping offered by VsCode (and cover most use cases)
- the only client of the graph atm is VsCode

I believe the graph code shouldn't look for its own configuration, but that it should be provided by the client instantiating the graph. In our case, settings are read by VsCode and given to the webview (see `dataviz.ts`), either via variable substitution or by sending a websocket message.
If you are familiar with React, I consider the graph to be a dumb component, not a smart one.
To clarify, this doesn't mean that `foam-core` would know about the graph, just `foam-vscode` (basically in `dataviz.ts`).

#### Structure

My only point would be to rename `nonExistingNote` with `placeholder`.

#### Tech approach

As the comment above doesn't include a tech approach to comment on will share my thoughts, happy to hear comments.
The color of the node is based on the `type` attribute of the node.
Current types are: note, nonExistingNote, unknown. In the future we could create nodes for other elements in the Foam model, like attachments and tags, which could be styled differently.
More broadly, I would like the type of a Foam note (and therefore of a graph node) to be exposed via a property in the YAML (or via a plugin).

Basically if a note has a `type: daily-note` YAML property, the graph node will have the same value.
That should allow us to customize the styling of the graph to our liking, and gives a lot of flexibility.

**Comment by jmg-duarte** [comment]
Since #438 was merged I'm closing this

**Comment by PSEUDO-SAPPHO** [comment]
can we select depth of backlink files-for backlinks of backlinks?  in graph or below the note.
in obsidian that shows in graph:
![image](https://user-images.githubusercontent.com/82666844/116128536-4bd2f880-a6de-11eb-9e7f-373ab3267dec.png)
